std/experimental/typecons.d

```diff
@@ -802,3 +802,224 @@ private template findCovariantFunction(alias finfo, Source, Fs...)
     else
         enum ptrdiff_t findCovariantFunction = x;
 }
+
+/**
+Type constructor for head-const variables.
+
+Head-const variables cannot be directly mutated or rebound, but references
+reached through the variable are typed with their original mutability.
+It is equivalent to $(D final) variables in D1 and Java, as well as
+$(D readonly) variables in C#.
+
+When $(D T) is a $(D const) or $(D immutable) type, $(D HeadConst) aliases
+to $(D T).
+*/
+struct HeadConst(T)
+    if (!is(T == const) && !is(T == immutable))
+{
+    import std.typecons : Proxy;
+
+    private T headConst_value;
+    mixin Proxy!headConst_value;
+
+    /**
+     * Construction is forwarded to the underlying type.
+     */
+    this(T other)
+    {
+        this.headConst_value = other;
+    }
+
+    /// Ditto
+    this(Args...)(auto ref Args args)
+        if (__traits(compiles, T(args)))
+    {
+        static assert((!is(T == struct) && !is(T == union)) || !isNested!T,
+            "Non-static nested type " ~ fullyQualifiedName!T ~ " must be " ~
+            "constructed explicitly at the call-site (e.g. auto s = " ~
+            "headConst(" ~ T.stringof ~ "(...));)");
+        this.headConst_value = T(args);
+    }
+
+    // Attaching function attributes gives less noisy error messages
+    pure nothrow @safe @nogc
+    {
+        /++
+         + All operators, including member access, are forwarded to the
+         + underlying value of type $(D T) except for these mutating operators,
+         + which are disabled.
+         +/
+        @disable void opAssign(Other)(Other other);
+        @disable void opOpAssign(string op, Other)(Other other); /// Ditto
+        @disable void opUnary(string op : "--")(); /// Ditto
+        @disable void opUnary(string op : "++")(); /// Ditto
+    }
+
+    /**
+     *
+     * $(D HeadConst!T) implicitly converts to an rvalue of type $(D T) through
+     * $(D AliasThis).
+     */
+    inout(T) headConst_get() inout
+    {
+        return headConst_value;
+    }
+
+    /// Ditto
+    alias headConst_get this;
+
+    /// Ditto
+    auto ref T opUnary(string op)()
```

Comment:
Any unary operator that could reasonably return by reference? Even if a UDT defines it as such, observing `ref` creates a hole in `HeadConst`. Let's go with value return.